protocols/kad/src/handler.rs

```diff
@@ -772,20 +770,41 @@ where
             ));
         }
 
-        if let Poll::Ready(Some(event)) = self.outbound_substreams.poll_next_unpin(cx) {
-            return Poll::Ready(event);
+        while let Poll::Ready(Some(event)) = self.outbound_substreams.poll_next_unpin(cx) {
+            if let ConnectionHandlerEvent::Custom(KademliaHandlerEvent::Available { substream }) =
+                event
+            {
+                self.reusable_outgoing_substreams.push(substream);
+            } else {
+                return Poll::Ready(event);
+            }
         }
 
         if let Poll::Ready(Some(event)) = self.inbound_substreams.poll_next_unpin(cx) {
             return Poll::Ready(event);
         }
 
-        let num_in_progress_outbound_substreams =
-            self.outbound_substreams.len() + self.num_requested_outbound_streams;
-        if num_in_progress_outbound_substreams < MAX_NUM_SUBSTREAMS {
-            if let Some(protocol) = self.requested_streams.pop_front() {
-                self.num_requested_outbound_streams += 1;
-                return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest { protocol });
+        while !self.pending_substream_requests.is_empty()
+            && self.outbound_substreams.len() + self.num_requested_outbound_streams
+                < MAX_NUM_SUBSTREAMS
+        {
+            if let Some((msg, user_data)) = self.pending_substream_requests.pop_front() {
+                if let Some(substream) = self.reusable_outgoing_substreams.pop() {
+                    self.outbound_substreams
+                        .push(OutboundSubstreamState::PendingSend(
+                            substream, msg, user_data,
+                        ));
+                    cx.waker().wake_by_ref();
+                } else {
+                    self.num_requested_outbound_streams += 1;
+                    let protocol = SubstreamProtocol::new(
+                        self.config.protocol_config.clone(),
+                        (msg, user_data),
+                    );
+                    return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                        protocol,
+                    });
+                }
```

Comment:
Like I mentioned in the other comment, I'd prefer if we just search `outbound_substreams` for one that is `Available` and have it mutate itself to `PendingSend` with the requested message.

We do something similar here somewhere already.

Reply:
We can do that unconditionally of the number of streams as a first step in here and if there are messages left, we can open new ones up to 32.

Let me know what you think!

Reply:
> We do something similar here somewhere already.

Yes, there is such pattern for incoming streams reuse already, there is an iteration there searching for streams waiting for subsequent incoming messages. But it felt less efficient searching things in `SelectAll` as I commented in the other place.